diff -Naur asterisk-1.8.orig/channels/chan_mspd.c asterisk-1.8/channels/chan_mspd.c
--- asterisk-1.8.orig/channels/chan_mspd.c	2014-05-07 12:19:34.000000000 +0300
+++ asterisk-1.8/channels/chan_mspd.c	2016-02-28 13:26:31.442522181 +0200
@@ -15,7 +15,7 @@
 
 #include "asterisk.h"
 
-//ASTERISK_FILE_VERSION(__FILE__, "$Revision: 1.94 $")
+//ASTERISK_FILE_VERSION(__FILE__, "$Revision: 1.95 $")
 
 #include <stdlib.h>
 #include <stdint.h>
@@ -1099,6 +1099,7 @@
     S8 cfg_bwe_master = -1;
     S8 cfg_bwe_high_band = -1;
     S8 cfg_bwe_low_band = -1;
+    S8 cfg_phone_band = -1;
     Boolean bwe_apply = False;
 
     MSPD_LOG(2, "Phone%d\n", mspd_global.n_phones + 1);
@@ -1208,6 +1209,9 @@
         else if (!strcasecmp(v->name,"wideband")) {
             phone->wideband = ast_true(v->value);
         }
+        else if(!strcasecmp(v->name, "phone_band")) {
+            cfg_phone_band = ast_true(v->value) ? True : False;
+        }
         else if(!strcasecmp(v->name, "bwe_master")) {
             cfg_bwe_master = ast_true(v->value) ? True : False;
         }
@@ -1222,6 +1226,10 @@
         }
     }
 
+    if(cfg_phone_band != -1) {
+        phone->wideband = cfg_phone_band;
+    }
+
     if(!phone->bwe.bwe_master) {
         bwe_apply = True;
     }
@@ -2001,7 +2009,7 @@
 {
     int i;
     int has_failed = 0;
-    int prev_val   = 0;
+   // int prev_val   = 0; unused
     char *cat = NULL;
     struct ast_config *cfg;
     mspd_ud_pref_t *prev_ud_pref = NULL;
@@ -2041,7 +2049,7 @@
 
     MSPD_LOCK (mspd_global.lock, goto FAIL);
 
-    prev_val = mspd_global.ud_prefs.ud_enable_supvsr;
+    //prev_val = mspd_global.ud_prefs.ud_enable_supvsr; unused variable
     for (i = 0; i < mspd_global.n_phones; ++i) {
         mspd_phone_t *phone = &mspd_global.phones[i];
 
diff -Naur asterisk-1.8.orig/channels/mspd/mspd_sti-lines.c asterisk-1.8/channels/mspd/mspd_sti-lines.c
--- asterisk-1.8.orig/channels/mspd/mspd_sti-lines.c	2014-05-07 12:19:35.000000000 +0300
+++ asterisk-1.8/channels/mspd/mspd_sti-lines.c	2016-02-28 13:25:26.573200953 +0200
@@ -1561,6 +1561,7 @@
         return;
     }
     line->is_timer_started = False;
+    line->is_flash_probe_on = False;
     *(struct TEMPO_EVENT_DESC **)value.sival_int = NULL;
 }
 
@@ -1806,6 +1807,7 @@
         sig.sigev_value.sival_int = (int) line->ptr_holder;
         sig.sigev_notify_function = onhook_event;
         line->is_timer_started = False;
+        line->is_flash_probe_on = False;
 
         int res = timer_create(CLOCK_REALTIME, &sig, &(line->flash_timer));
         if (res < 0)
