--- a/setup.py
+++ b/setup.py
@@ -214,7 +214,10 @@ class PyBuildExt(build_ext):
             # If a module has already been built statically,
             # don't build it here
             if ext.name in sys.builtin_module_names:
-                self.extensions.remove(ext)
+                if not cross_compiling:
+                    self.extensions.remove(ext)
+                else:
+                    print(ext.name, "in sys.builtin_module_names, building anyway")
 
         # Parse Modules/Setup and Modules/Setup.local to figure out which
         # modules are turned on in the file.
